node/network/statement-distribution/src/lib.rs

```diff
@@ -310,33 +304,62 @@ impl PeerData {
 }
 
 // A statement stored while a relay chain head is active.
-//
-// These are orderable first by (Seconded, Valid, Invalid), then by the underlying hash,
-// and lastly by the signing validator's index.
-#[derive(PartialEq, Eq)]
+#[derive(Debug)]
 struct StoredStatement {
-	compact: CompactStatement,
+	comparator: StoredStatementComparator,
 	statement: SignedFullStatement,
```

Comment:
`comparator.compact` can always be constructed from `statement`. I think I can see the rationale for why you set things up like this, but even so it feels like this design can be improved. In particular, I'd worry about someone modifying the `StoredStatement` for some (valid) reason, but letting `statement` and `comparator.compact` get out of sync.

My instinct would be to get rid of the `StoredStatementComparator` type and manually impl `PartialEq` for `StoredStatement`.

Reply:
For `HashSet::get`, you need a `Borrow` implementation, which requires that we store the `comparator`, not construct it ad-hoc.

`StoredStatement` is primarily passed around as an immutable reference, so I think the mutation concern, although technically right, is not something I perceive as a risk at all. Especially because the issue would only be caused if `HashSet` had a `get_mut` function, which it doesn't, and for logical reasons, will probably never get. However, even if `HashSet` had `get_mut`, and we started to pass around mutable `StoredStatements` for some reason, then we would need someone to change the comparator of the stored statement. Which is itself a very strange action that I can't think of any reasonable motivation for. I think it is very unlikely that any of those things will happen in our world, much less all 3.